snuba/util.py

```diff
@@ -102,64 +167,39 @@ def column_expr(column_name, body, alias=None, aggregate=None):
         expr = escape_col(column_name)
 
     if aggregate:
-        if expr:
-            expr = u'{}({})'.format(aggregate, expr)
-            if aggregate == 'uniq':  # default uniq() result to 0, not null
-                expr = 'ifNull({}, 0)'.format(expr)
-        else:  # This is the "count()" case where the '()' is already provided
-            expr = aggregate
+        expr = function_expr(aggregate, expr)
 
     alias = escape_col(alias or column_name)
 
     return alias_expr(expr, alias, body)
 
 
 def complex_column_expr(expr, body, depth=0):
-    # TODO instead of the mutual recursion between column_expr and complex_column_expr
-    # we should probably encapsulate all this logic in a single recursive column_expr
-    if depth == 0:
-        # we know the first item is a function
-        ret = expr[0]
-        assert SAFE_FUNCTION_RE.match(ret)
-        expr = expr[1:]
-
-        # if the last item of the toplevel is a string, it's an alias
-        alias = None
-        if len(expr) > 1 and isinstance(expr[-1], six.string_types):
-            alias = expr[-1]
-            expr = expr[:-1]
-    else:
-        # is this a nested function call?
-        if len(expr) > 1 and isinstance(expr[1], tuple):
-            ret = expr[0]
-            assert SAFE_FUNCTION_RE.match(ret)
-            expr = expr[1:]
+    function_tuple = is_function(expr, depth)
+    if function_tuple is None:
+        raise ValueError('complex_column_expr was given an expr %s that is not a function at depth %d.' % (expr, depth))
+
+    name, args, alias = function_tuple
+    out = []
+    i = 0
+    while i < len(args):
+        next_2 = args[i:i+2]
+        if is_function(next_2, depth+1):
```

Comment:
I am not sure I get which is the case where the first two arguments resulting from is_function(expr, depth) (line 182) could represent themselves a function. Isn't a function always expressed as

> [name, [params]]

Reply:
No. we haven't done that historically. It would be nice if that changed though. see examples like: https://github.com/getsentry/snuba/blob/master/tests/test_util.py#L213

In the tests.

Reply:
ok